### PR TITLE
micro buffer internal lib

### PR DESCRIPTION
**Pull request by pahumadad**
Basic Buffer lib with no support for concurrency

### CODE REVIEW

File: src/buffer.cc [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        Impl(const std::size_t size);
+
+        void push(std::unique_ptr<Message> msg);
+        std::unique_ptr<Message> pop();
+    };
+}
+
+// Implementation
+micro::Buffer::Impl::Impl(const std::size_t size)
+    : queue(std::deque<std::unique_ptr<Message>>())
```
Comment (fortizc) [inline src/buffer.cc]: This not needed. Here you are copy (or moving in the best case) the deque. Simply remove this line

File: include/micro/core/buffer.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        bool is_full() const;
+        std::size_t size() const;
+        std::size_t capacity() const;
+
+        void push(std::unique_ptr<Message> msg);
+        std::unique_ptr<Message> pop();
+
+    private:
+        class Impl;
+        std::unique_ptr<Impl> impl;
```
Comment (fortizc) [inline include/micro/core/buffer.h]: Normally this memeber is called `pimpl`
Impl -> **Impl**emntation
pimpl -> **p**ointer to **impl**ementation

File: src/buffer.cc [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+std::size_t micro::Buffer::capacity() const
+{
+    return impl->max_size;
+}
+
+void micro::Buffer::push(std::unique_ptr<Message> msg)
+{
+    return impl->push(std::move(msg));
```
Comment (fortizc) [inline src/buffer.cc]: This return must be removed

File: include/micro/core/buffer.h (hunk 2 above)
Comment (pahumadad) [reply include/micro/core/buffer.h]: renamed! de17ad1b4a5fac1516264559f955cbb9de1a35cd

File: src/buffer.cc (hunk 3 above)
Comment (pahumadad) [reply src/buffer.cc]: what a shame! 🤦‍♂ 
cda5549150d370f63dc07d404636b24cab5bfeeb

File: src/buffer.cc (hunk 1 above)
Comment (pahumadad) [reply src/buffer.cc]: you're totally right! 41a566913c4cf417a08384d195c1ffea780a86ba

**Review by fortizc — APPROVED** [review]
LGTM

**Pull request merged**